### PR TITLE
ChaCha20-Poly1305: use EVP_CIPHER_CTX_encrypting instead of ctx->encrypt

### PR DESCRIPTION
**Pull request by ronaldtse**
During #4564 I found that in `crypto/evp/e_chacha20_poly1305.c` that the calls to check whether the context was encrypting were done using `ctx-encrypt` instead of the EVP `EVP_CIPHER_CTX_encrypting` function.

After this replacement, no more cipher functions directly call `ctx->encrypt`.

This is a contribution from [Ribose Inc.](https://github.com/riboseinc).

##### Checklist

### CODE REVIEW

**Comment by ronaldtse** [comment]
Thanks @paulidale 😉

**Comment by ronaldtse** [comment]
@levitte removal of that line caused a number of undefined references...

File: crypto/evp/e_chacha20_poly1305.c [hunk 1]
```diff
@@ -290,12 +290,12 @@ static int chacha20_poly1305_cipher(EVP_CIPHER_CTX *ctx, unsigned char *out,
 
             Poly1305_Update(POLY1305_ctx(actx), temp, POLY1305_BLOCK_SIZE);
         }
-        Poly1305_Final(POLY1305_ctx(actx), ctx->encrypt ? actx->tag
+        Poly1305_Final(POLY1305_ctx(actx), EVP_CIPHER_CTX_encrypting(ctx) ? actx->tag
```
Comment (levitte) [inline crypto/evp/e_chacha20_poly1305.c]: This line goes past 80 columns, needs being broken up differently.  I suggest breaking the line after the comma.

File: crypto/evp/e_chacha20_poly1305.c [hunk 2]
```diff
@@ -232,7 +232,7 @@ static int chacha20_poly1305_cipher(EVP_CIPHER_CTX *ctx, unsigned char *out,
             else if (len != plen + POLY1305_BLOCK_SIZE)
                 return -1;
 
-            if (ctx->encrypt) {                 /* plaintext */
+            if (EVP_CIPHER_CTX_encrypting(ctx)) {                 /* plaintext */
```
Comment (levitte) [inline crypto/evp/e_chacha20_poly1305.c]: Reindent the comment, I suggest reducing all spaces between `{` and the comment to one space

File: crypto/evp/e_chacha20_poly1305.c [hunk 3]
```diff
@@ -383,7 +383,7 @@ static int chacha20_poly1305_ctrl(EVP_CIPHER_CTX *ctx, int type, int arg,
         return 1;
 
     case EVP_CTRL_AEAD_GET_TAG:
-        if (arg <= 0 || arg > POLY1305_BLOCK_SIZE || !ctx->encrypt)
+        if (arg <= 0 || arg > POLY1305_BLOCK_SIZE || !EVP_CIPHER_CTX_encrypting(ctx))
```
Comment (levitte) [inline crypto/evp/e_chacha20_poly1305.c]: Line longer than 80 columns.  I suggest breaking it before the last `||`

**Comment by dot-asm** [comment]
This was not actually a coincidence. The EVP_CIPHER_CTX_encrypting is effectively for those callers who has no access to internal EVP structure, i.e. those outside libcrypto. But for EVP code *itself* there are only downsides in using it. Well, I'm with mind damaged by assembly programming judge rather generated code than source. But even at source level it doesn't really give you an edge [as compiler is perfectly capable of catching misconduct].

**Comment by levitte** [comment]
> @levitte removal of that line caused a number of undefined references...

Ok.  I assumed that the goal would be to make `e_chacha20_poly1305.c` less dependent on EVP internals, otherwise I don't quite see the point with this exercise.

The rest of the stuff is about direct reference to `cipher_data`, where there are `EVP_CIPHER_CTX_get_cipher_data` and `EVP_CIPHER_CTX_set_cipher_data`...

**Comment by dot-asm** [comment]
> This was not actually a coincidence.

Just in case, choice of `ctx->encrypt` was not a coincidence.

**Comment by levitte** [comment]
I recall that we at some point discussed moving `crypto/evp/e_*.c` to their respective algos, so using the API instead of accessing internals directly would be a helpful move in that direction.

Of course, for combos like chacha20/poly1305, it's difficult to figure where to move it...  so maybe @dot-asm is right about this one.  I'm afraid that makes this PR a futile exercise :-/

**Comment by ronaldtse** [comment]
@levitte the latest commit gets rid of the evp_locl.h include.

File: crypto/evp/e_chacha20_poly1305.c [hunk 4]
```diff
@@ -316,7 +317,8 @@ static int chacha20_poly1305_cleanup(EVP_CIPHER_CTX *ctx)
 {
     EVP_CHACHA_AEAD_CTX *actx = aead_data(ctx);
     if (actx)
-        OPENSSL_cleanse(ctx->cipher_data, sizeof(*actx) + Poly1305_ctx_size());
+        OPENSSL_cleanse(EVP_CIPHER_CTX_get_cipher_data(ctx),
+            POLY1305_actx_size(actx));
```
Comment (levitte) [inline crypto/evp/e_chacha20_poly1305.c]: Nit: align to containing parenthesis

File: crypto/evp/e_chacha20_poly1305.c [hunk 5]
```diff
@@ -328,8 +330,9 @@ static int chacha20_poly1305_ctrl(EVP_CIPHER_CTX *ctx, int type, int arg,
     switch(type) {
     case EVP_CTRL_INIT:
         if (actx == NULL)
-            actx = ctx->cipher_data
-                 = OPENSSL_zalloc(sizeof(*actx) + Poly1305_ctx_size());
+            EVP_CIPHER_CTX_set_cipher_data(ctx,
+                OPENSSL_zalloc(POLY1305_actx_size(actx)));
+            actx = EVP_CIPHER_CTX_get_cipher_data(ctx);
```
Comment (levitte) [inline crypto/evp/e_chacha20_poly1305.c]: May I suggest this instead?

``` C
            actx = OPENSSL_zalloc(sizeof(*actx) + Poly1305_ctx_size());
            EVP_CIPHER_CTX_set_cipher_data(ctx, actx);
```

Looks a bit cleaner, doesn't it?

File: crypto/evp/e_chacha20_poly1305.c [hunk 6]
```diff
@@ -347,9 +350,10 @@ static int chacha20_poly1305_ctrl(EVP_CIPHER_CTX *ctx, int type, int arg,
         if (actx) {
             EVP_CIPHER_CTX *dst = (EVP_CIPHER_CTX *)ptr;
 
-            dst->cipher_data =
-                   OPENSSL_memdup(actx, sizeof(*actx) + Poly1305_ctx_size());
-            if (dst->cipher_data == NULL) {
+            EVP_CIPHER_CTX_set_cipher_data(dst,
+                    OPENSSL_memdup(actx, POLY1305_actx_size(actx)));
+
+            if (EVP_CIPHER_CTX_get_cipher_data(dst) == NULL) {
```
Comment (levitte) [inline crypto/evp/e_chacha20_poly1305.c]: Because you will end up with ugly indentation, I suggest this instead:

``` C
            actx = OPENSSL_memdup(actx, sizeof(*actx) + Poly1305_ctx_size());
            EVP_CIPHER_CTX_set_cipher_data(dst, actx);

            if (actx == NULL) {
```

(either way, you changed the size to `OPENSSL_memdup`)

Comment (ronaldtse) [reply crypto/evp/e_chacha20_poly1305.c]: Addressed.

File: crypto/evp/e_chacha20_poly1305.c (hunk 5 above)
Comment (ronaldtse) [reply crypto/evp/e_chacha20_poly1305.c]: Thank you for the great solution 👍

File: crypto/evp/e_chacha20_poly1305.c (hunk 4 above)
Comment (ronaldtse) [reply crypto/evp/e_chacha20_poly1305.c]: Done.

File: crypto/evp/e_chacha20_poly1305.c [hunk 7]
```diff
@@ -14,9 +14,9 @@
 
 # include <openssl/evp.h>
 # include <openssl/objects.h>
-# include "evp_locl.h"
 # include "internal/evp_int.h"
 # include "internal/chacha.h"
+# include "internal/poly1305.h"
```
Comment (levitte) [inline crypto/evp/e_chacha20_poly1305.c]: Why is this needed?

Comment (ronaldtse) [reply crypto/evp/e_chacha20_poly1305.c]: There is a reference to the Poly block size.

Comment (levitte) [reply crypto/evp/e_chacha20_poly1305.c]: Yes?  Thing is, there is already an inclusion of `internal/poly1305.h`.  See line 147.

Comment (ronaldtse) [reply crypto/evp/e_chacha20_poly1305.c]: Sorry I must be going crazy... removed.

Comment (levitte) [reply crypto/evp/e_chacha20_poly1305.c]: You're human, like the rest of us ;-)

Comment (ronaldtse) [reply crypto/evp/e_chacha20_poly1305.c]: Haha, as human as I can be 😉

**Comment by levitte** [comment]
@dot-asm, things have continued here.  I'm unsure if your concerns went away or if I should consider your comments a blocker...

**Comment by dot-asm** [comment]
Once again, choice of `ctx->encrypt` was not a coincidence/accident/unconscious. As [assembly] implementations are getting faster, execution overhead in surrounding code is taking larger proportion of time. Modern processors might be mighty beasts, but they are still sensitive to various resources' depletion: cache trashing, branch prediction logic, translation lookaside buffers. You won't notice difference on larger blocks, and maybe not on all systems, but it's there. For example I measured ~5% difference on Sandy Bridge and Cortex-A53 on shortest inputs. And this is in benchmark. We should recognize that benchmarks tend to exhibit favorable results, because they effectively "worm up" subsystems. This means that in real life overheads would be larger. Unfortunately they are hard to measure till they take "astronomic" proportions. And for this reason it's only appropriate to adhere to programming techniques that attempt to minimize pressure on hardware resources. Avoiding excessive calls is one of such techniques. And so it minimizing branches. [Yes, it means that I'd argue against excessive calls to evp_lib.c even if EVP modules are moved to other directories. As long as you're within libcrypto boundary, it's only appropriate to rely on internal headers such evp_locl.h.]

**Comment by levitte** [comment]
@ronaldtse, while I generally would do the exact thing as you, I also think @dot-asm's argument against this PR are sound.

@dot-asm, that last part within [ ] is actually an excellent argument *against* moving the `e_*.c` files away from `crypto/evp/`.  No, I don't think it's appropriate to include `evp_locl.h` from anywhere but files in `crypto/evp/`...  if we do, we might as well move that one to `crypto/include/internal/`.

**Comment by dot-asm** [comment]
> I don't think it's appropriate to include evp_locl.h from anywhere but files in crypto/evp/... if we do, we might as well move that one to crypto/include/internal/.

Of course it would be hardly appropriate to include specifically crypto/evp/evp_locl.h and it would be better moved. But we are free to do that at any point if so decided.

**Comment by ronaldtse** [comment]
@dot-asm @levitte it's certainly your call and the reason seems reasonable. However this does beg the question then why is all other code in `crypto/evp/*` only using `EVP_CIPHER_CTX_*` instead of `ctx->*`. If there is already a 5% gain for eliminating 1 level of function calls...

**Comment by dot-asm** [comment]
It's in human nature to concentrate on one thing at a time, but it has a side effect. People are then prone to fail to take into consideration *all* the [relevant/necessary] details, which in turn might result to suboptimal decisions. [Humans are also limited in their capacities and might fail to notice in time for modification (which can be short), or collect energy to do something about it later...]

Just in case, 5% is not a universal measure, it's just an example in *this* case, and even in this case it's system-specific. The only thing one can be sure of is that coefficient would be larger with faster implementations... [And of course smaller/negligible with faster implementations...]
